### PR TITLE
[Core] Add some Device.InvokeOnMainThread helpers

### PR DESCRIPTION
**Pull request by knocte**
These helpers keep popping up on forums/stackoverflow[1][2],
would be nice to include them out of the box.

[1] https://forums.xamarin.com/discussion/51720/wait-for-device-begininvokeonmainthread
[2] https://stackoverflow.com/a/47941859

### CODE REVIEW

**Comment by kvpt** [comment]
In this case I think that the **Begin** prefix is confusing because with **BeginInvoke** you just launch the action, it's a fire forget, but in this case, after the await of the call the action is terminated.

**Comment by knocte** [comment]
@kvpt thanks, renamed

**Comment by knocte** [comment]
@StephaneDelcroix fixed thanks

File: Xamarin.Forms.Core/Device.cs [hunk 1]
```diff
@@ -104,6 +104,85 @@ public static void BeginInvokeOnMainThread(Action action)
 			PlatformServices.BeginInvokeOnMainThread(action);
 		}
 
+		public static Task InvokeOnMainThreadAsync(Action action)
+		{
+			var tcs = new TaskCompletionSource<object>();
```
Comment (paymicro) [inline Xamarin.Forms.Core/Device.cs]: Perhaps `action` can reuse `Func` code
```C#
return InvokeOnMainThreadAsync(() => { action(); return true; });
```

File: Xamarin.Forms.Core/Device.cs [hunk 2]
```diff
@@ -104,6 +104,85 @@ public static void BeginInvokeOnMainThread(Action action)
...
+				catch (Exception e)
+				{
+					tcs.SetException(e);
+				}
+			});
+
+			return tcs.Task;
+		}
+
+		public static Task InvokeOnMainThreadAsync(Task actionTask)
```
Comment (campersau) [inline Xamarin.Forms.Core/Device.cs]: The task is only awaited on the main thread but not started on it. Is this the intended behaviour?

**Comment by samhouts** [comment]
build

**Comment by knocte** [comment]
> Perhaps `action` can reuse `Func` code

Done, thanks.

> The task is only awaited on the main thread but not started on it. Is this the intended behaviour?

Oh damn, that's a good point. I copied the implementation from here: https://stackoverflow.com/a/47941859 ; and didn't realise about it. I asked the author of the answer a question around this, and the answer doesn't sound very convincing to me. I guess I should remove the two overloads that receive Task objects? Thoughts?

**Comment by AmrAlSayed0** [comment]
[xamarin/Essentials/Xamarin.Essentials/MainThread](https://github.com/xamarin/Essentials/tree/master/Xamarin.Essentials/MainThread) There are some here but they don't cover all of what is in this PR.

**Comment by knocte** [comment]
> There are some here

I only see 2 methods there, same ones as we already have in Xamarin.Forms, no asynchronous API.

**Comment by andreinitescu** [comment]
In my opinion, any new implementation which creates Task should make sure it works properly when Activity is recreated, see #2801

**Comment by knocte** [comment]
> any new implementation

Not really new implementations, just wrappers around what's already there.

**Comment by knocte** [comment]
> Oh damn, that's a good point. I copied the implementation from here: 
> https://stackoverflow.com/a/47941859 ; and didn't realise about it. I asked the author of the answer
> a question around this, and the answer doesn't sound very convincing to me. I guess I should
> remove the two overloads that receive Task objects? 

Replaced the overloads that received Task objects, with overloads that receive Func<Task> objects,
so that the action/function that generates the task is executed on the main thread. I think this should be good now.

**Comment by knocte** [comment]
(Sorry, didn't mean to close. Just reopened.)

File: Xamarin.Forms.Core/Device.cs [hunk 3]
```diff
@@ -104,6 +104,65 @@ public static void BeginInvokeOnMainThread(Action action)
...
+		{
+			return InvokeOnMainThreadAsync(funcTask);
+		}
+
+		public static async Task<SynchronizationContext> GetMainThreadSynchronizationContextAsync()
+		{
+			SynchronizationContext ret = null;
+			await InvokeOnMainThreadAsync(() =>
+				ret = SynchronizationContext.Current
+			);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Device.cs]:
```suggestion
			).ConfigureAwait(false);
```

**Comment by rmarinho** [comment]
Hey @knocte  can you rebase please ? thanks

**Comment by knocte** [comment]
Why is a rebase needed if there are no conflicts?

**Comment by StephaneDelcroix** [comment]
build

**Comment by knocte** [comment]
BTW I cannot inspect what's wrong with CI, the "OSX Phase" links me to somewhere that ends up showing a 401 error.

**Comment by StephaneDelcroix** [comment]
@knocte I'm rebuilding to figure that out

**Comment by StephaneDelcroix** [comment]
@knocte the build issue on the CI is some provisioning problem on iOS. that's why @rmarinho requested a rebase. It'd be nice if you could do it.

**Comment by knocte** [comment]
rebased

**Comment by StephaneDelcroix** [comment]
build

**Comment by StephaneDelcroix** [comment]
@rmarinho please figure out what's still wrong with provisioning and rebased PR.

File: Xamarin.Forms.Core/Device.cs [hunk 4]
```diff
@@ -104,6 +104,65 @@ public static void BeginInvokeOnMainThread(Action action)
...
+					}
+				}
+			);
+
+			return tcs.Task;
+		}
+
+		public static Task InvokeOnMainThreadAsync(Func<Task> funcTask)
+		{
+			return InvokeOnMainThreadAsync(funcTask);
```
Comment (campersau) [inline Xamarin.Forms.Core/Device.cs]: This calls itself. Do something like this:

```cs
Func<Task<object>> dummyFunc = () => { funcTask(); return null; };
return InvokeOnMainThreadAsync(dummyFunc);
```

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Device.cs]: good catch

Comment (knocte) [reply Xamarin.Forms.Core/Device.cs]: Thanks, fixed.

**Pull request merged**